Store agent group data

RamosFe opened the pull request:
|Related issue|
|---|
[16069](https://github.com/wazuh/wazuh/issues/16069)



## Description

Added `Agent groups` and `Agent groups full` status data to the `MasterHandler` and changed log of the `cluster_control.py` script.

## Example 

```
Cluster name: wazuh

Connected nodes (1):

    master-node (wazuh-master)
        Version: 4.4.1
        Type: master
        Active agents: 0

    worker (172.18.0.4)
        Version: 4.4.1
        Type: worker
        Active agents: 1
        Status:
            Last keep Alive:
                Last received: 2023-03-16T11:58:05.985666Z.
            Integrity check:
                Last integrity check: 0.007s (2023-03-16T11:58:13.274594Z - 2023-03-16T11:58:13.281172Z).
                Permission to check integrity: True.
            Integrity sync:
                Last integrity synchronization: n/a (n/a - n/a).
                Synchronized files: Shared: 0 | Missing: 0 | Extra: 0.
            Agents-info:
                Last synchronization: 0.001s (2023-03-16T11:58:16.613794Z - 2023-03-16T11:58:16.615173Z).
                Number of synchronized chunks: 1.
                Permission to synchronize agent-info: True.
            Agents-groups:
                Last synchronization: 0.001s (2023-03-16T11:58:16.822680Z - 2023-03-16T11:58:16.823920Z).
                Number of synchronized chunks: 1.
            Agents-groups full:
                Last synchronization: n/a (n/a - n/a).
                Number of synchronized chunks: 0.
```

Selutario reviewed (changes requested):
The changes look fine but there are some missing things:

#### 1. cluster_control CLI

`Agent-groups` and `Agent-groups full` tasks should also be printed when `more` option is not used:
```
# /var/ossec/bin/cluster_control -i 
Cluster name: wazuh

Last completed synchronization for connected nodes (2):
    worker1 (192.168.32.3): Integrity check: 2023-03-24T13:37:25.352068Z | Integrity sync: 2023-03-24T13:34:51.828252Z | Agents-info: n/a | Last keep alive: 2023-03-24T13:36:42.743341Z.
    worker2 (192.168.32.5): Integrity check: 2023-03-24T13:37:25.338433Z | Integrity sync: 2023-03-24T13:34:51.831001Z | Agents-info: n/a | Last keep alive: 2023-03-24T13:36:42.761959Z.
```

#### 2. Error in cluster log

This error is being logged in the master's `cluster.log`:
```
2023/03/24 13:35:33 ERROR: [Worker worker1] [Main] Unhandled error processing request 'b'syn_w_g_e'': unsupported operand type(s) for -: 'float' and 'str'
Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.4.0-py3.9.egg/wazuh/core/cluster/common.py", line 799, in dispatch
    command, payload = self.process_request(command, payload)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.4.0-py3.9.egg/wazuh/core/cluster/master.py", line 265, in process_request
    return c_common.end_sending_agent_information(logger, start_time, data.decode())
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.4.0-py3.9.egg/wazuh/core/cluster/common.py", line 1425, in end_sending_agent_information
    msg = f"Finished in {(get_utc_now().timestamp() - start_time):.3f}s. Updated {data['updated_chunks']} chunks."
TypeError: unsupported operand type(s) for -: 'float' and 'str'
```

RamosFe commented:
## Update 
Added msg when more option is not used
````
# var/ossec/bin/cluster_control -i
Cluster name: wazuh

Last completed synchronization for connected nodes (1):
    worker (172.18.0.4): Integrity check: 2023-03-28T15:34:38.801440Z | Integrity sync: 2023-03-28T15:33:53.499310Z | Agents-info: 2023-03-28T15:34:16.094609Z | Last keep alive: 2023-03-28T15:33:53.379383Z | Agent-groups: 2023-03-28T15:34:33.444899Z | Agent-groups full: n/a.
````

Selutario commented on framework/wazuh/core/cluster/master.py:
```diff
@@ -650,7 +657,13 @@ async def send_entire_agent_groups_information(self):
+        # Updates Agent groups full status
+        self.send_full_agent_groups_status['date_start'] = start_time.strftime(DECIMALS_DATE_FORMAT)
+        self.send_full_agent_groups_status['date_end'] = end_time.strftime(DECIMALS_DATE_FORMAT)
+        self.send_agent_groups_status['n_synced_chunks'] = len(local_agent_groups_information)
```
```suggestion
        self.send_full_agent_groups_status['n_synced_chunks'] = len(local_agent_groups_information)
```

RamosFe replied:
Fixed:

https://github.com/wazuh/wazuh/blob/04dd591be4ddb5842b4df5a3575498e7612ddc60/framework/wazuh/core/cluster/master.py#L666

Selutario commented on framework/scripts/cluster_control.py:
```diff
@@ -120,7 +120,9 @@ def calculate_seconds(start_time, end_time):
                         f"Integrity sync: {node_info['status']['last_sync_integrity']['date_end_master']} | " \
                         f"Agents-info: {node_info['status']['last_sync_agentinfo']['date_end_master']} | " \
-                        f"Last keep alive: {node_info['status']['last_keep_alive']}.\n"
+                        f"Last keep alive: {node_info['status']['last_keep_alive']} | " \
+                        f"Agent-groups: {node_info['status']['last_sync_agentgroup']['date_end']} | " \
+                        f"Agent-groups full: {node_info['status']['last_sync_full_agentgroup']['date_end']}.\n"
```
I think it would be better to place these two tasks before `Last keep alive` (just a suggestion):
```suggestion
                        f"Agent-groups: {node_info['status']['last_sync_agentgroup']['date_end']} | " \
                        f"Agent-groups full: {node_info['status']['last_sync_full_agentgroup']['date_end']} | " \
                        f"Last keep alive: {node_info['status']['last_keep_alive']}.\n"
```

RamosFe replied:
Changed the format as requested:

```
# var/ossec/bin/cluster_control -i
Cluster name: wazuh

Last completed synchronization for connected nodes (1):
    worker (172.18.0.2): Integrity check: 2023-03-29T14:23:16.620492Z | Integrity sync: 2023-03-29T14:23:07.574395Z | Agents-info: n/a | Agent-groups: 2023-03-29T14:23:19.333603Z | Agent-groups full: n/a | Last keep alive: 2023-03-29T14:22:49.035152Z.
```

Selutario reviewed (changes requested):
Everything looks good now! :tada: There are however two things to update that I forgot to mention:
1. **API spec:** The `GET /cluster/healthcheck` response is now different so it should be updated ([here](https://github.com/wazuh/wazuh/blob/04dd591be4ddb5842b4df5a3575498e7612ddc60/api/api/spec/spec.yaml#L1977-L2048) and [here](https://github.com/wazuh/wazuh/blob/04dd591be4ddb5842b4df5a3575498e7612ddc60/api/api/spec/spec.yaml#L9593-L9660)).
2. **Documentation:** It will be required to update the examples in [this documentation page](https://documentation.wazuh.com/current/user-manual/reference/tools/cluster-control.html).

RamosFe commented:
# Update

Updated documentation: https://github.com/wazuh/wazuh-documentation/pull/6009
Updated spec files: https://github.com/wazuh/wazuh/pull/16394/commits/dad739056ef9c3053964c5a09e58659b206b6aee
